_posts/2025-04-28-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,429 @@
+<4> Use GitHub authorization code flow to support the login endpoint with a dedicated Quarkus OIDC `login` https://quarkus.io/guides/security-openid-connect-multitenancy[tenant] configuration.
+<5> Use the client id and secret generated in the <<register-github-application>> section.
+
+[NOTE]
+====
+Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and do not interfere with dev mode. 
+====
+
+=== Implement Login endpoint
+
+Currently, MCP clients can not use the authorization code flow themselves. Therefore, we implement an OAuth2 login endpoint that will return a GitHub token for the user to use with MCP clients, which can work with bearer tokens.
+
+Add another dependency: to support Qute templates:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-rest-qute</artifactId> <1>
+</dependency>
+----
+<1> `quarkus-rest-qute` is required to generate HTML pages. Its version is defined in the Quarkus BOM.
+
+and implement the login endpoint:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.oidc.AccessTokenCredential;
+import io.quarkus.oidc.UserInfo;
+import io.quarkus.qute.Template;
+import io.quarkus.qute.TemplateInstance;
+import io.quarkus.security.Authenticated;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+
+@Path("/login")
+@Authenticated
+public class LoginResource {
+
+    @Inject
+    UserInfo userInfo; <1>
+    
+    @Inject
+    AccessTokenCredential accessToken; <2>
+
+    @Inject
+    Template accessTokenPage;
+
+    @GET
+    @Produces("text/html")
+    public TemplateInstance poem() {
+        return accessTokenPage
+           .data("name", userInfo.getName())
+           .data("accessToken", accessToken.getToken()); <3>
+    }
+}
+----
+<1> GitHub access tokens are binary and Quarkus OIDC indirectly verifies them by using them to request GutHub specific `UserInfo` representation.
+<2> `AccessTokenCredential` is used to capture a binary GitHub access token.
+<3> After the user logs in to GitHub and is redirected to this endpoint, the access token will be returned to the user in an HTML page generated with a simple https://github.com/quarkiverse/quarkus-mcp-server/tree/main/samples/secure-mcp-sse-server/src/main/resources/templates/accessTokenPage.html[Qute template].
+Of course, you would not do that in a real application. It is just an example to demonstrate the capability.
+
+=== Package and run the MCP Server
+
+Package the MCP server application:
+
+[source,shell]
+----
+mvn package
+----
+
+Run it:
+
+[source,shell]
+----
+java -jar target/quarkus-app/quarkus-run.jar
+----
+
+[NOTE]
+====
+You can also run the MCP server with `jbang`.
+
+You have to set the `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:
```

Comment:
```suggestion
When using Maven coordinates its advantageous to use `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:
```

that setting is not needed for jbang.

you can also just run the generated .jar directly with java or jbang.  so adjusted to the correct thing above.

Reply:
actually - you sholdn't even needed to use uber-jar since you are not using @fatjar thus dependencies will be resolved anyway. So its unnecessary - its mvn install that is needed.

Reply:
@maxandersen Indeed, thanks for the hint, I just dropped a ref to the uber jar property for now.